storage/innobase/dict/dict0dict.cc

```diff
@@ -6781,3 +6781,178 @@ dict_tf_to_row_format_string(
+
+	ut_ad(name);
+	ut_ad(data);
+
+	rw_lock_x_lock(&dict_operation_lock);
+	dict_mutex_enter_for_mysql();
+
+	trx = trx_allocate_for_background();
+	trx->op_info = "insert zip_dict";
+	trx->dict_operation_lock_mode = RW_X_LATCH;
+	trx_start_if_not_started(trx);
+
+	err = dict_create_add_zip_dict(name, name_len, data, data_len, trx);
+
+	if(err == DB_SUCCESS)
+	{
+		trx_commit_for_mysql(trx);
+	}
+	else
+	{
+		trx->op_info = "rollback of internal trx on zip_dict table";
+		trx_rollback_to_savepoint(trx, NULL);
+		ut_a(trx->error_state == DB_SUCCESS);
+	}
+	trx->op_info = "";
+	trx->dict_operation_lock_mode = 0;
+	trx_free_for_background(trx);
+
+	dict_mutex_exit_for_mysql();
+	rw_lock_x_unlock(&dict_operation_lock);
+
+	return err;
+}
+/********************************************************************//**
+Get single compression dictionary id for the given
+(table id, column pos) pair.
+@return	DB_SUCCESS if OK, DB_RECORD_NOT_FOUND if not found */
+UNIV_INTERN
+dberr_t
+dict_get_dictionary_id_by_key(
+/*================================*/
+	ulint	table_id,	/*!< in: table id */
+	ulint	column_pos,	/*!< in: column position */
+	ulint*	dict_id)	/*!< out: zip_dict id */
+{
+	dberr_t		err = DB_SUCCESS;
+	trx_t*		trx;
+
+	rw_lock_x_lock(&dict_operation_lock);
+	dict_mutex_enter_for_mysql();
+
+	trx = trx_allocate_for_background();
+	trx->op_info = "get zip dict id by composite key";
+	trx->dict_operation_lock_mode = RW_X_LATCH;
+	/* TODO: (ZIP_DICT) set read-only mode for transaction */
+	trx_start_if_not_started(trx);
+
+	err = dict_create_get_zip_dict_id_by_reference(table_id, column_pos, dict_id, trx);
+
+	if(err == DB_SUCCESS)
+	{
+		trx_commit_for_mysql(trx);
+	}
+	else
+	{
+		trx->op_info = "rollback of internal trx on zip_dict_cols table";
+		trx_rollback_to_savepoint(trx, NULL);
+		ut_a(trx->error_state == DB_SUCCESS);
+	}
+	trx->op_info = "";
+	trx->dict_operation_lock_mode = 0;
+	trx_free_for_background(trx);
+
+	dict_mutex_exit_for_mysql();
+	rw_lock_x_unlock(&dict_operation_lock);
+
+	return err;
+}
+/********************************************************************//**
+Get compression dictionary info (name and data) for the given id.
+Allocates memory in name->str and data->str on success.
+Must be freed with mem_free().
+@return	DB_SUCCESS if OK, DB_RECORD_NOT_FOUND if not found */
+UNIV_INTERN
+dberr_t
+dict_get_dictionary_info_by_id(
+/*================================*/
+	ulint	dict_id,	/*!< in: table name */
+	char**	name,		/*!< out: dictionary name */
+	ulint*	name_len,	/*!< out: dictionary name length*/
+	char**	data,		/*!< out: dictionary data */
+	ulint*	data_len)	/*!< out: dictionary data length*/
+{
+	dberr_t		err = DB_SUCCESS;
+	trx_t*		trx;
+
+	rw_lock_x_lock(&dict_operation_lock);
+	dict_mutex_enter_for_mysql();
+
+	trx = trx_allocate_for_background();
+	trx->op_info = "get zip dict name and data by id";
+	trx->dict_operation_lock_mode = RW_X_LATCH;
+	/* TODO: (ZIP_DICT) set read-only mode for transaction */
```

Comment:
Changed to `rw_lock_s_lock(&dict_operation_lock)` and `trx->dict_operation_lock_mode = RW_S_LATCH` for now. As there is no proper way to set read-only mode for background transactions, `trx` here is still read-write. This function can be rewritten with InnoDB API routines to read data from `SYS_ZIP_DICT*` tables in future.